client/client.go

```diff
@@ -79,26 +89,34 @@ func (c *Client) Dial(server string) (*gokeyless.Conn, error) {
 		return nil, err
 	}
 
+	c.m.Lock()
+	defer c.m.Unlock()
 	c.conns[server] = gokeyless.NewConn(conn)
 	return c.conns[server], nil
 }
 
 // DialAny smartly chooses one of the keyless servers given. (Opting to reuse an existing connection if possible)
 func (c *Client) DialAny(ski gokeyless.SKI) (*gokeyless.Conn, error) {
-	servers := c.allServers[ski]
+	servers := c.getServers(ski)
 	if len(servers) == 0 {
-		return nil, errors.New("no servers given")
+		return nil, fmt.Errorf("no servers registered for SKI %02x", ski)
 	}
 
-	var existing []*gokeyless.Conn
 	for _, server := range servers {
-		if conn, ok := c.conns[server]; ok {
-			existing = append(existing, conn)
+		c.m.RLock()
+		conn, ok := c.conns[server]
+		c.m.RUnlock()
+		if ok {
+			if conn.Use() {
+				return conn, nil
+			}
+			c.m.Lock()
+			if c.conns[server] == conn {
```

Comment:
This could be a loop that breaks when `c.conns[server] == conn` (and `!conn.Use()`)